### PR TITLE
[Feature](explode) support explode map type

### PR DESCRIPTION
**Pull request by amorynan**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by amorynan** [comment]
run buildall

**Comment by amorynan** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java [hunk 1]
```diff
@@ -1929,6 +1930,15 @@ private void initTableFunction() {
         addTableFunctionWithCombinator(EXPLODE, Type.WILDCARD_DECIMAL, Function.NullableMode.ALWAYS_NULLABLE,
                 Lists.newArrayList(new ArrayType(Type.WILDCARD_DECIMAL)), false,
                 "_ZN5doris19DummyTableFunctions7explodeEPN9doris_udf15FunctionContextERKNS1_13CollectionValE");
+
+        initTableFunctionListWithCombinator(EXPLODE_MAP);
```
Comment (BiteTheDDDDt) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java:1934]: Are those code useless?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/generator/ExplodeMapOuter.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            throw new AnalysisException("only support map type for explode_map function but got "
+                    + child().getDataType());
+        }
+    }
+
+    @Override
+    public List<FunctionSignature> getSignatures() {
+        return ImmutableList.of(
+                FunctionSignature.ret(new StructType(ImmutableList.of(
+                                new StructField("col1", ((MapType) child().getDataType()).getKeyType(), true, ""),
```
[the remark this answers is not in the document]
Comment (amorynan) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/generator/ExplodeMapOuter.java:70]: yes, this is struct default col name

File: be/src/vec/exprs/table_function/vexplode_map.h [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+#include "common/status.h"
+#include "vec/columns/column_map.h"
+#include "vec/data_types/data_type.h"
+#include "vec/data_types/data_type_array.h"
+#include "vec/data_types/data_type_map.h"
+#include "vec/exprs/table_function/table_function.h"
+#include "vec/functions/array/function_array_utils.h"
+
+namespace doris {
+namespace vectorized {
```
[the remark this answers is not in the document]
Comment (HappenLee) [reply be/src/vec/exprs/table_function/vexplode_map.h:31]: shoud change like `namespace doris::vectorized {`

File: be/src/vec/exprs/table_function/vexplode_map.h [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+class ColumnMapExecutionData {
+public:
+    void reset() {
+        map_nullmap_data = nullptr;
+        map_col = nullptr;
+        offsets_ptr = nullptr;
+    }
+
+public:
```
Comment (HappenLee) [inline be/src/vec/exprs/table_function/vexplode_map.h:46]: here seems should be `private` ?
and shoud `_map_nullmap_data`

Comment (amorynan) [reply be/src/vec/exprs/table_function/vexplode_map.h:46]: maybe not use _map_nullmap_data

File: fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java (hunk 1 above)
Comment (amorynan) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java:1934]: it's usefull

**Comment by amorynan** [comment]
run buildall

**Comment by amorynan** [comment]
run buildall

**Comment by amorynan** [comment]
run pipelinex_p0

File: regression-test/suites/nereids_p0/sql_functions/table_function/explode_map.groovy [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+    // insert values
+    sql """ insert into sdu values (0, "zhangsan", {"Chinese":"80","Math":"60","English":"90"}); """
+    sql """ insert into sdu values (1, "lisi", {"null":null}); """
+    sql """ insert into sdu values (2, "wangwu", {"Chinese":"88","Math":"90","English":"96"}); """
+    sql """ insert into sdu values (3, "lisi2", {null:null}); """
+    sql """ insert into sdu values (4, "amory", NULL); """
+
+    qt_sql """ select * from sdu order by id; """
+    qt_explode_sql """ select name, k,v from sdu lateral view explode_map(score) tmp as k,v order by id;"""
```
Comment (yiguolei) [inline regression-test/suites/nereids_p0/sql_functions/table_function/explode_map.groovy:43]: id is not in select list， why not throw exception?

Comment (amorynan) [reply regression-test/suites/nereids_p0/sql_functions/table_function/explode_map.groovy:43]: lateral view explode is table function, which has all column info , and this is same action with explode too.

**Comment by amorynan** [comment]
run pipelinex_p0

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/generator/ExplodeMapOuter.java (hunk 2 above)
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/generator/ExplodeMapOuter.java:70]: create a function in structType like: `StructType constructStructType(List<DataType> fieldTypes)` to generate default column name struct type. and call this function here and in `org.apache.doris.nereids.trees.expressions.literal.StructLiteral#computeDataType`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindExpression.java [hunk 6]
```diff
@@ -624,8 +631,34 @@ protected boolean condition(Rule rule, Plan plan) {
...
+                    LogicalGenerate ret = new LogicalGenerate<>(
+                            boundFunctionGenerators, slotBuilder.build(), generate.child());
+                    if (expandAlias.size() > 0) {
+                        // we need a project to deal with explode(map) to struct with field alias
+                        // project should contains: generator.child slot + expandAlias
+                        List<NamedExpression> allProjectSlots = generate.child().getOutput().stream()
+                                .map(NamedExpression.class::cast)
+                                .collect(Collectors.toList());
+                        allProjectSlots.addAll(expandAlias);
+                        return new LogicalProject<>(allProjectSlots, ret);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindExpression.java:659]:
```suggestion
                        ret = new LogicalProject<>(allProjectSlots, ret);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindExpression.java [hunk 7]
```diff
@@ -587,7 +593,7 @@ protected boolean condition(Rule rule, Plan plan) {
                     // we need to do cast before set operation, because we maybe use these slot to do shuffle
                     // so, we must cast it before shuffle to get correct hash code.
                     List<List<NamedExpression>> childrenProjections = setOperation.collectChildrenProjections();
-                    ImmutableList.Builder<List<SlotReference>> childrenOutputs = ImmutableList.builder();
+                    Builder<List<SlotReference>> childrenOutputs = ImmutableList.builder();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindExpression.java:596]: do not refactor other code in this PR

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 8]
```diff
@@ -1021,14 +1021,21 @@ protected LogicalPlan withGenerate(LogicalPlan plan, LateralViewContext ctx) {
             return plan;
         }
         String generateName = ctx.tableName.getText();
-        String columnName = ctx.columnName.getText();
+        // if later view explode map type, we need to add a project to convert map to struct
+        String columnName = ctx.columnNames.get(0).getText();
+        List<String> expandColumnNames = Lists.newArrayList();
+        if (ctx.columnNames.size() > 1) {
+            columnName = "expand_cols";
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:1028]: use column name generator to generate column name when connect context not null.
```java
ConnectContext.get().getStatementContext().generateColumnName()
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindExpression.java [hunk 9]
```diff
@@ -624,8 +631,34 @@ protected boolean condition(Rule rule, Plan plan) {
                         Slot boundSlot = new SlotReference(slot.getNameParts().get(1), generator.getDataType(),
                                 generator.nullable(), ImmutableList.of(slot.getNameParts().get(0)));
                         slotBuilder.add(boundSlot);
+                        // the boundSlot may has two situation:
+                        // 1. the expandColumnsAlias is not empty, we should use make boundSlot expand to multi alias
+                        // 2. the expandColumnsAlias is empty, we should use origin boundSlot
+                        if (generate.getExpandColumnAlias() != null
+                                && !CollectionUtils.isEmpty(generate.getExpandColumnAlias().get(i))) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindExpression.java:638]: `generate.getExpandColumnAlias().get(i)` could generate OutOfBound exception, should check first

**Comment by amorynan** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Review by yiguolei — APPROVED** [review]

**Pull request merged**